File: test_plugin/ios/AdjustTesting.m
//
//  AdjustSdk.h
//  Adjust SDK
//
//  Created by Abdullah Obaied (@obaied) on 25th October 2016.
//  Copyright © 2012-2018 Adjust GmbH. All rights reserved.
//

#import "AdjustTesting.h"
#import "ATLTestLibrary.h"
#import "AdjustTestingCommandListener.h"

@implementation AdjustTesting {
    AdjustTestingCommandListener *adjustCommandListener;
    ATLTestLibrary *testLibrary;
    NSMutableArray *selectedTests;
    NSMutableArray *selectedTestDirs;
}

RCT_EXPORT_MODULE(AdjustTesting);

#pragma mark - Public methods

RCT_EXPORT_METHOD(startTestSession:(NSString *)baseUrl) {
    NSLog(@"startTestSession():");
    if (![self isFieldValid:baseUrl]) {
        return;
    }

    adjustCommandListener = [[AdjustTestingCommandListener alloc] init];
    
    testLibrary = [ATLTestLibrary testLibraryWithBaseUrl:baseUrl
                                           andCommandDelegate:adjustCommandListener];
    
    for (id object in selectedTests) {
        [testLibrary addTest:object];
    }
    
    for (id object in selectedTestDirs) {
        [testLibrary addTestDirectory:object];
    }
    
    [testLibrary startTestSession:@"react_native4.12.3@ios4.12.3"];
}

RCT_EXPORT_METHOD(addInfoToSend:(NSString *)key value:(NSString *)value) {
    if (testLibrary != nil) {
        NSLog(@"addInfoToSend(): with key %@ and %@", key, value);
        [testLibrary addInfoToSend:key value:value];
    }
}

RCT_EXPORT_METHOD(sendInfoToServer:(NSString *)basePath) {
    if (testLibrary != nil) {
        NSLog(@"sendInfoToServer(): with basePath %@", basePath);
        [testLibrary sendInfoToServer:basePath];
    }
}

RCT_EXPORT_METHOD(addTest:(NSString *)testToAdd) {
    NSLog(@"addtest(): %@", testToAdd);
    
    if (selectedTests == nil) {
        selectedTests = [NSMutableArray array];
    }
    
    [selectedTests addObject:testToAdd];
}

RCT_EXPORT_METHOD(addTestDirectory:(NSString *)testDirToAdd) {
    NSLog(@"addtestDirectory(): %@", testDirToAdd);
    
    if (selectedTestDirs == nil) {
        selectedTestDirs = [NSMutableArray array];
    }
    
    [selectedTestDirs addObject:testDirToAdd];
}

#pragma mark - Private & helper methods

- (BOOL)isFieldValid:(NSObject *)field {
    if (field == nil) {
        return false;
    }
    
    // Check if its an instance of the singleton NSNull
    if ([field isKindOfClass:[NSNull class]]) {
        return false;
    }
    
    // If `field` can be converted to a string, check if it has any content.
    NSString *str = [NSString stringWithFormat:@"%@", field];
    if (str != nil) {
        if ([str length] == 0) {
            return false;
        }
        
        if ([str isEqualToString:@"null"]) {
            return false;
        }
    }
    
    return true;
}

@end

File: test/lib/android/src/main/java/com/adjust/testingmodule/AdjustTesting.java
//
//  AdjustPackage.java
//  Adjust
//
//  Created by Abdullah Obaied on 2018-02-23.
//  Copyright (c) 2018 Adjust GmbH. All rights reserved.
//  See the file MIT-LICENSE for copying permission.
//

package com.adjust.testingmodule;

import com.adjust.testlibrary.TestLibrary;

import android.util.Log;
import java.util.ArrayList;
import java.util.List;

import com.facebook.react.bridge.*;
import com.facebook.react.modules.core.*;

public class AdjustTesting extends ReactContextBaseJavaModule {
    private static final String TAG = "AdjustTesting";

    private TestLibrary testLibrary;
    private List<String> selectedTests = new ArrayList<String>();
    private List<String> selectedTestDirs = new ArrayList<String>();

    public AdjustTesting(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "AdjustTesting";
    }

    @ReactMethod
    public void startTestSession(String baseUrl) {
        Log.d(TAG, "startTestSession() with baseUrl[" + baseUrl + "]");

        testLibrary = new TestLibrary(baseUrl, new CommandListener(getReactApplicationContext()));

        for(int i = 0; i < selectedTests.size(); i++) {
            testLibrary.addTest(selectedTests.get(i));
        }

        for(int i = 0; i < selectedTestDirs.size(); i++) {
            testLibrary.addTestDirectory(selectedTestDirs.get(i));
        }

        testLibrary.startTestSession("react_native4.13.0@android4.13.0");
    }

    @ReactMethod
    public void addInfoToSend(String key, String value) {
        if (testLibrary != null) {
            Log.d(TAG, "AddInfoToSend() with key[" + key + "] and value[" + value + "]");
            testLibrary.addInfoToSend(key, value);
        }
    }

    @ReactMethod
    public void sendInfoToServer(String basePath) {
        Log.d(TAG, "sendInfoToServer(): " + basePath);
        if (testLibrary != null) {
            testLibrary.sendInfoToServer(basePath);
        }
    }

    @ReactMethod
    public void addTest(String test) {
        this.selectedTests.add(test);
    }

    @ReactMethod
    public void addTestDirectory(String testDir) {
        this.selectedTestDirs.add(testDir);
    }
}
